src/mds/MDSRank.cc

```diff
@@ -3054,11 +3054,77 @@ bool MDSRankDispatcher::handle_command(
 
     damage_table.erase(id);
     return true;
+  } else if (prefix == "drop cache") {
+    command_drop_cache(ds);
+    return true;
   } else {
     return false;
   }
 }
 
+void MDSRank::command_drop_cache(stringstream *ds)
+{
+  if (not is_active()) {
+    dout(5) << __func__ << ": MDS not active, no-op" << dendl;
+    return;
+  }
+
+  if (not mdcache->trim(UINT64_MAX))
+    return;
+
+  server->recall_client_state(1, true);
+  _wait_for_client_recall_state_acks();
+
+  std::unique_ptr<Formatter> fmtr(new JSONFormatter(true));
+
+  std::stringstream ss;
+  fmtr.get()->open_object_section("result");
+
+  int retval = _command_flush_journal(&ss);
+  fmtr.get()->open_object_section("flush_journal");
+  fmtr.get()->dump_string("message", ss.str());
+  fmtr.get()->dump_int("return_code", retval);
+  fmtr.get()->close_section();
+
+  retval = mdcache->cache_status(fmtr.get());
+  if (retval != 0)
+   *ds << "Failed to get cache status: " << cpp_strerror(retval);
+
+  fmtr.get()->close_section();
+  fmtr->flush(*ds);
+}
+
+bool MDSRank::_wait_for_client_recall_state_acks()
+{
+  const unsigned int TIMEOUT = g_conf->get_val<uint64_t>("mds_session_recall_"
+							"ack_timeout");
+  set<Session*> unacked_sessions;
+  mono_time time_at_beg = mono_clock::now();
+  std::chrono::duration<double> time_passed = mono_clock::now() - time_at_beg;
+
+  sessionmap.get_client_session_set(unacked_sessions);
+
+  while (time_passed.count() < TIMEOUT) {
```

Comment:
The MDS should do a wait, not an infinite loop. You need to add a waiter interface for this.

Reply:
> The MDS should do a wait, not an infinite loop.

It's not infinite. The looping stops on timeout.

Reply:
Sorry my wording was wrong. It's not infinite but it **is** a busy-wait. i.e. the CPU loops until the condition is true... for `TIMEOUT` seconds. This isn't acceptable.

Reply:
Got it.